File: walk_tracker.js
// walk_tracker.js
#!/usr/bin/env node
const fs = require('fs');
const { program } = require('commander');

const CONFIG_FILE = 'walk_config.json';
const DATA_FILE = 'walks.json';

let weight = 70.0;
let walks = [];

function loadConfig() {
    if (fs.existsSync(CONFIG_FILE)) {
        try {
            const cfg = JSON.parse(fs.readFileSync(CONFIG_FILE));
            weight = cfg.weight || 70.0;
        } catch (e) {}
    }
}

function saveConfig() {
    fs.writeFileSync(CONFIG_FILE, JSON.stringify({ weight }));
}

function loadWalks() {
    if (fs.existsSync(DATA_FILE)) {
        try {
            walks = JSON.parse(fs.readFileSync(DATA_FILE));
        } catch (e) {}
    }
}

function saveWalks() {
    fs.writeFileSync(DATA_FILE, JSON.stringify(walks, null, 2));
}

function addWalk(dist, dur, w) {
    if (w === undefined) w = weight;
    const speed = dist / (dur / 60.0);
    let met;
    if (speed < 3.0) met = 2.0;
    else if (speed < 5.0) met = 3.0;
    else if (speed < 6.5) met = 3.8;
    else if (speed < 8.0) met = 5.0;
    else met = 6.0;
    const cal = met * w * (dur / 60.0);
    const entry = {
        date: new Date().toISOString(),
        distance: dist,
        duration: dur,
        speed,
        weight: w,
        calories: Math.round(cal * 10) / 10
    };
    walks.push(entry);
    saveWalks();
    console.log(`✅ Walk added: ${dist} km in ${dur} min, ${entry.calories} kcal`);
}

function listWalks() {
    if (!walks.length) {
        console.log('No walks recorded.');
        return;
    }
    console.log('\n📋 Walks:');
    walks.forEach((w, i) => {
        const dt = w.date.slice(0, 16).replace('T', ' ');
        console.log(`${i+1}. ${dt} | ${w.distance} km | ${w.duration} min | ${w.speed.toFixed(1)} km/h | ${w.calories} kcal`);
    });
}

function stats() {
    if (!walks.length) {
        console.log('No walks yet.');
        return;
    }
    let totalDist = 0, totalDur = 0, totalCal = 0;
    for (const w of walks) {
        totalDist += w.distance;
        totalDur += w.duration;
        totalCal += w.calories;
    }
    console.log('\n🚶 Walk Summary');
    console.log(`Total walks: ${walks.length}`);
    console.log(`Total distance: ${totalDist.toFixed(2)} km`);
    console.log(`Total duration: ${totalDur} min (${(totalDur/60).toFixed(1)} h)`);
    console.log(`Total calories: ${totalCal.toFixed(0)} kcal`);
}

function exportCSV(filename) {
    const header = 'Date,Distance (km),Duration (min),Speed (km/h),Weight (kg),Calories (kcal)\n';
    const rows = walks.map(w =>
        `${w.date},${w.distance},${w.duration},${w.speed.toFixed(1)},${w.weight},${w.calories}`
    ).join('\n');
    fs.writeFileSync(filename, header + rows);
    console.log(`Exported ${walks.length} walks to ${filename}`);
}

program
    .command('weight <kg>')
    .action((kg) => {
        weight = parseFloat(kg);
        saveConfig();
        console.log(`Default weight set to ${weight} kg`);
    });

program
    .command('add <distance> <duration>')
    .option('--weight <kg>', 'Weight for this walk')
    .action((distance, duration, options) => {
        const w = options.weight ? parseFloat(options.weight) : undefined;
        addWalk(parseFloat(distance), parseFloat(duration), w);
    });

program
    .command('list')
    .action(listWalks);

program
    .command('stats')
    .action(stats);

program
    .command('export [filename]')
    .action((filename = 'walks.csv') => exportCSV(filename));

program.parse(process.argv);

File: WalkTracker.cs
// WalkTracker.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

class Config
{
    [JsonPropertyName("weight")]
    public double Weight { get; set; } = 70.0;
}

class Walk
{
    [JsonPropertyName("date")]
    public string Date { get; set; }
    [JsonPropertyName("distance")]
    public double Distance { get; set; }
    [JsonPropertyName("duration")]
    public double Duration { get; set; }
    [JsonPropertyName("speed")]
    public double Speed { get; set; }
    [JsonPropertyName("weight")]
    public double Weight { get; set; }
    [JsonPropertyName("calories")]
    public double Calories { get; set; }
}

class WalkTracker
{
    private static readonly string ConfigFile = "walk_config.json";
    private static readonly string DataFile = "walks.json";
    private static double weight = 70.0;
    private static List<Walk> walks = new List<Walk>();
    private static readonly JsonSerializerOptions options = new JsonSerializerOptions { WriteIndented = true };

    static void LoadConfig()
    {
        if (!File.Exists(ConfigFile)) return;
        string json = File.ReadAllText(ConfigFile);
        Config cfg = JsonSerializer.Deserialize<Config>(json);
        if (cfg != null) weight = cfg.Weight;
    }

    static void SaveConfig()
    {
        Config cfg = new Config { Weight = weight };
        string json = JsonSerializer.Serialize(cfg, options);
        File.WriteAllText(ConfigFile, json);
    }

    static void LoadWalks()
    {
        if (!File.Exists(DataFile)) return;
        string json = File.ReadAllText(DataFile);
        walks = JsonSerializer.Deserialize<List<Walk>>(json) ?? new List<Walk>();
    }

    static void SaveWalks()
    {
        string json = JsonSerializer.Serialize(walks, options);
        File.WriteAllText(DataFile, json);
    }

    static void AddWalk(double distance, double duration, double? weightOverride)
    {
        double w = weightOverride ?? weight;
        double speed = distance / (duration / 60.0);
        double met;
        if (speed < 3.0) met = 2.0;
        else if (speed < 5.0) met = 3.0;
        else if (speed < 6.5) met = 3.8;
        else if (speed < 8.0) met = 5.0;
        else met = 6.0;
        double cal = met * w * (duration / 60.0);
        Walk entry = new Walk
        {
            Date = DateTime.Now.ToString("o"),
            Distance = distance,
            Duration = duration,
            Speed = speed,
            Weight = w,
            Calories = Math.Round(cal, 1)
        };
        walks.Add(entry);
        SaveWalks();
        Console.WriteLine($"✅ Walk added: {distance} km in {duration} min, {entry.Calories} kcal");
    }

    static void ListWalks()
    {
        if (!walks.Any())
        {
            Console.WriteLine("No walks recorded.");
            return;
        }
        Console.WriteLine("\n📋 Walks:");
        for (int i = 0; i < walks.Count; i++)
        {
            var w = walks[i];
            string dt = w.Date.Substring(0, 16).Replace('T', ' ');
            Console.WriteLine($"{i+1}. {dt} | {w.Distance} km | {w.Duration} min | {w.Speed:F1} km/h | {w.Calories} kcal");
        }
    }

    static void Stats()
    {
        if (!walks.Any())
        {
            Console.WriteLine("No walks yet.");
            return;
        }
        double totalDist = walks.Sum(w => w.Distance);
        double totalDur = walks.Sum(w => w.Duration);
        double totalCal = walks.Sum(w => w.Calories);
        Console.WriteLine("\n🚶 Walk Summary");
        Console.WriteLine($"Total walks: {walks.Count}");
        Console.WriteLine($"Total distance: {totalDist:F2} km");
        Console.WriteLine($"Total duration: {totalDur} min ({totalDur/60:F1} h)");
        Console.WriteLine($"Total calories: {totalCal:F0} kcal");
    }

    static void ExportCSV(string filename)
    {
        using var writer = new StreamWriter(filename);
        writer.WriteLine("Date,Distance (km),Duration (min),Speed (km/h),Weight (kg),Calories (kcal)");
        foreach (var w in walks)
        {
            writer.WriteLine($"{w.Date},{w.Distance},{w.Duration},{w.Speed:F1},{w.Weight},{w.Calories}");
        }
        Console.WriteLine($"Exported {walks.Count} walks to {filename}");
    }

    static void Main(string[] args)
    {
        LoadConfig();
        LoadWalks();
        if (args.Length == 0)
        {
            Console.WriteLine("Usage: WalkTracker [weight|add|list|stats|export]");
            return;
        }
        string cmd = args[0];
        switch (cmd)
        {
            case "weight":
                if (args.Length < 2) { Console.WriteLine("Usage: weight <kg>"); return; }
                weight = double.Parse(args[1]);
                SaveConfig();
                Console.WriteLine($"Default weight set to {weight} kg");
                break;
            case "add":
                if (args.Length < 3) { Console.WriteLine("Usage: add <distance> <duration> [--weight <kg>]"); return; }
                double dist = double.Parse(args[1]);
                double dur = double.Parse(args[2]);
                double? w = null;
                for (int i = 3; i < args.Length; i++)
                {
                    if (args[i] == "--weight" && i+1 < args.Length)
                    {
                        w = double.Parse(args[i+1]);
                        i++;
                    }
                }
                AddWalk(dist, dur, w);
                break;
            case "list":
                ListWalks();
                break;
            case "stats":
                Stats();
                break;
            case "export":
                string filename = args.Length > 1 ? args[1] : "walks.csv";
                ExportCSV(filename);
                break;
            default:
                Console.WriteLine("Unknown command");
                break;
        }
    }
}
